SeaPkg/Tools/GenSeaArtifacts/gen_aux/src/report.rs

```diff
@@ -0,0 +1,321 @@
+            rule_coverable: format!("{:#x}", uncovered + covered_by_rule),
+            rule_covered: format!("{:#x}", covered_by_rule),
+            rule_covered_percent: format!("{:.2}%", coverable_cov_percent),
+            segments,
+        })
+    }
+
+    /// Writes the report to a file
+    pub fn to_file(&self, path: std::path::PathBuf) -> anyhow::Result<()> {
+        let mut file = std::fs::File::create(path)?;
+        let buffer = serde_json::to_vec_pretty(&self)?;
+        file.write_all(&buffer)?;
+        Ok(())
+    }
+}
+
+/// A wrapper around a list of segments that allows for inserting new segments into the list.
+struct SegmentList {
+    size_of_image: u32,
+    segments: Vec<Segment>,
+}
+
+impl SegmentList {
+    /// Creates a new segment list with a single segment that spans the entire image.
+    fn new(size: u32) -> Self {
+        SegmentList {
+            size_of_image: size,
+            segments: vec![Segment::new(0, size, false, "".to_string())],
+        }
+    }
+
+    /// Consumes the list and returns the inner vector of segments.
+    pub fn into_inner(self) -> Vec<Segment> {
+        self.segments
+    }
+
+    /// Adds segments to the report based off the PE image.
+    pub fn add_segments_from_image(&mut self, image: &[u8]) -> Result<()> {
+        self.add_ro_sections(image)?;
+        self.add_section_padding(image)?;
+        self.add_pe_header(image)?;
+        Ok(())
+    }
+
+    /// Adds a section (as one or more segments) if the section is read-only, including the padding at the end of the section.
+    fn add_ro_sections(&mut self, image: &[u8]) -> Result<()> {
+        let pe = goblin::pe::PE::parse(image)?;
+        let sections = pe.sections;
+
+        fn read(characteristics: u32) -> bool {
+            characteristics & 0x40000000 != 0
+        }
+
+        fn write(characteristics: u32) -> bool {
+            characteristics & 0x80000000 != 0
+        }
+
+        fn execute(characteristics: u32) -> bool {
+            characteristics & 0x20000000 != 0
+        }
+
+        for i in 0..sections.len() {
+            let section = &sections[i];
+            let next_section_start = if i + 1 < sections.len() {
+                sections[i + 1].virtual_address
+            } else {
+                self.size_of_image
+            };
+
+            let c = section.characteristics;
+            match (read(c), write(c), execute(c)) {
+                (true, false, false) => {
+                    self.insert(Segment::new(section.virtual_address, next_section_start, true, "ReadOnly Section".to_string()))?
+                },
+                (true, false, true) => {
+                    self.insert(Segment::new(section.virtual_address, next_section_start, true, "ReadExecute Section".to_string()))?
+                }
+                _ => continue,
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Finds the padding at the end of each section and adds it as a covered segment.

```

Comment:
I think it would make more sense to remove the padding from the overall image size instead of adding it to the section size.  This would give us a more accurate covered% as padding will increase our average.

Reply:
This is actually one of the reasons why I provided two calculations:

1. All coverable: `(PE header + covered sections + padding + rules) / Total size`
2. Only coverable by rules: `rules / (rules) + uncovered)`

I'm completely fine with updating the final calculation for (1) to be `(PE Header + covered sections + rules) / (Total Size - padding)` If we think it is a more accurate view of the data, but I think keeping the padding segments in the raw data is valuable

Reply:
No other requests for change. Closing.